gnucash/gnome-utils/gnc-autoclear.cpp

```diff
@@ -0,0 +1,274 @@
+    gboolean debugging_enabled = qof_log_check (G_LOG_DOMAIN, QOF_LOG_DEBUG);
+    GQuark autoclear_quark = g_quark_from_static_string ("autoclear");
+
+    g_return_val_if_fail (GNC_IS_ACCOUNT (account), FALSE);
+    g_return_val_if_fail (splits != nullptr, FALSE);
+
+    empty_vec = {};
+    *splits = nullptr;
+
+    /* Extract which splits are not cleared and compute the amount we have to clear */
+    for (GList *node = xaccAccountGetSplitList (account); node; node = node->next)
+    {
+        auto split = static_cast<Split*> (node->data);
+
+        if (xaccSplitGetReconcile (split) != NREC)
+            toclear_value = gnc_numeric_sub_fixed
+                (toclear_value, xaccSplitGetAmount (split));
+        else if (gnc_numeric_zero_p (xaccSplitGetAmount (split)))
+            DEBUG ("skipping zero-amount split %p", split);
+        else if (gnc_numeric_check (xaccSplitGetAmount (split)))
+            DEBUG ("skipping invalid-amount split %p", split);
+        else if (end_date != INT64_MAX &&
+                 xaccTransGetDate (xaccSplitGetParent (split)) > end_date)
+            DEBUG ("skipping split after statement_date %p", split);
+        else
+            nc_vector.emplace_back (split);
+    }
+
+    if (gnc_numeric_zero_p (toclear_value))
+    {
+        g_set_error (error, autoclear_quark, AUTOCLEAR_NOP,
+                     _("Account is already at Auto-Clear Balance."));
+        goto skip_knapsack;
+    }
+    else if (nc_vector.empty ())
+    {
+        g_set_error (error, autoclear_quark, AUTOCLEAR_NOP,
+                     _("No uncleared splits found."));
+        goto skip_knapsack;
+    }
+
+    start_ticks = next_update_tick = clock ();
+
+    for (auto& split : nc_vector)
+    {
+        auto amount = xaccSplitGetAmount (split);
+        std::vector<WorkItem> workvector {};
+
+        workvector.reserve (sack.size () + 1);
+        if (sack.find (amount.num) != sack.end ())
+            workvector.emplace_back (amount.num, static_cast<SplitVec>(1));
+        else
+        {
+            SplitVec new_splits = { split };
+            workvector.emplace_back (amount.num, new_splits);
+        }
+
+        // printf ("new split. sack size = %ld\n", sack.size());
+
+        for (auto& [map_value, map_splits] : sack)
+        {
+            gint64 new_value;
+            if (__builtin_add_overflow (map_value, amount.num, &new_value))
+            {
+                g_set_error (error, autoclear_quark, AUTOCLEAR_OVERLOAD,
+                             "Overflow error: Amount numbers are too large!");
+                goto skip_knapsack;
+            }
+            if (map_splits == static_cast<SplitVec>(1) || sack.find (new_value) != sack.end ())
+                workvector.emplace_back (new_value, static_cast<SplitVec>(1));
+            else
+            {
+                auto new_splits = map_splits;
+                new_splits.emplace_back (split);
+                workvector.emplace_back (new_value, new_splits);
```

Comment:
```
new_splits.push_back(split);
```
`emplace_back` allocates and constructs a copy of  and stores the pointer to that in the vector. You don't need or want the extra layer of indirection. `push_back` copies the pointer into the slot in the split.